File: apps/desktop/src/renderer/services/llmProviders.ts
import { api } from "@/services/api";
import type { components } from "@/types/api.generated";

/**
 * BYOK 多服务商配置（设置·模型配置）数据层。
 *
 * 一个账号可配置一组 OpenAI 兼容服务商（各自 label + key + 端点）。
 * 账号默认「模型组合」走 `/v1/users/me/llm-model-profiles`（见
 * {@link import("@/services/llmModelProfiles")}）。
 * REST 类型由后端 OpenAPI 生成（仓库根 `pnpm gen:types`）。
 */

type Schemas = components["schemas"];

/** 单个 BYOK 服务商的设置视图（永不含明文 Key）。 */
export type LlmProviderView = Schemas["LlmProviderView"];

/**
 * 设置·模型配置的服务商列表 + 部署级能力。
 *
 * 部署级字段（`billing_mode` / `platform_available` / `platform_model` /
 * `default_model_profile_id`）描述账号/部署，不属于任何单个服务商。
 */
export type LlmProvidersResponse = Schemas["LlmProvidersResponse"];

/** 新增服务商入参。 */
export type CreateLlmProviderInput = Schemas["CreateLlmProviderRequest"];
/** 部分更新服务商入参（省略 api_key 保留已存密文）。 */
export type UpdateLlmProviderInput = Schemas["UpdateLlmProviderRequest"];

/** 服务商列表 + 部署能力（设置页 BYOK 区数据源）。 */
export function listLlmProviders(): Promise<LlmProvidersResponse> {
  return api.get<LlmProvidersResponse>("/v1/users/me/llm-providers");
}

/** 新增一个 OpenAI 兼容服务商（Key 加密存储，状态置 'unchecked'）。 */
export function createLlmProvider(
  input: CreateLlmProviderInput,
): Promise<LlmProviderView> {
  return api.post<LlmProviderView>("/v1/users/me/llm-providers", input);
}

/** 更新某服务商（端点 / label；api_key 省略则保留已存 Key）。 */
export function updateLlmProvider(
  providerId: string,
  input: UpdateLlmProviderInput,
): Promise<LlmProviderView> {
  return api.patch<LlmProviderView>(
    `/v1/users/me/llm-providers/${providerId}`,
    input,
  );
}

/** 删除某服务商（组合槽位引用由后端静默回落，不硬失败）。 */
export function deleteLlmProvider(
  providerId: string,
): Promise<{ status: string }> {
  return api.delete<{ status: string }>(
    `/v1/users/me/llm-providers/${providerId}`,
  );
}

/** 探测某服务商端点，持久化并返回 'active' / 'error' + supports_tools。 */
export function testLlmProvider(providerId: string): Promise<LlmProviderView> {
  return api.post<LlmProviderView>(
    `/v1/users/me/llm-providers/${providerId}/test`,
  );
}

/**
 * 账号「有效聊天模型」所在服务商的工具调用支持位——喂给工具软门禁
 * （{@link import("@/lib/llmToolsGate").needsToolsGateHint}）。
 * `providerId` 通常来自 `ModelCatalogResponse.current.provider_id`（默认组合展开后的主槽）；
 * 平台 / keyless 时无 id → undefined（不提示）。
 */
export function defaultChatSupportsTools(
  response: LlmProvidersResponse | undefined | null,
  providerId?: string | null,
): boolean | null | undefined {
  const id = providerId?.trim();
  if (!id) return undefined;
  return response?.providers.find((p) => p.id === id)?.supports_tools;
}

File: apps/server/agentcore/api/routes/llm_providers.py
"""BYOK LLM provider management (设置·模型配置 · 多服务商列表).

A user's list of OpenAI-compatible providers: list (+ deployment caps), add, edit,
remove, and connectivity-test each. Account default combination is under
``/users/me/llm-model-profiles``.
"""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from agentcore.api.dependencies import AuthUser, get_db
from agentcore.api.schemas import (
    CreateLlmProviderRequest,
    LlmProvidersResponse,
    LlmProviderView,
    StatusResponse,
    UpdateLlmProviderRequest,
)
from agentcore.core.logging import get_logger
from agentcore.llm.provider_service import LlmProviderService, LlmProvidersView
from agentcore.llm.provider_service import LlmProviderView as ServiceProviderView

logger = get_logger(__name__)

router = APIRouter(prefix="/users/me/llm-providers", tags=["llm-providers"])


def get_llm_provider_service(session: AsyncSession = Depends(get_db)) -> LlmProviderService:
    return LlmProviderService(session)


def _provider_to_response(view: ServiceProviderView) -> LlmProviderView:
    return LlmProviderView(
        id=view.id,
        label=view.label,
        base_url=view.base_url,
        status=view.status,
        masked_key=view.masked_key,
        supports_tools=view.supports_tools,
        message=view.message,
        created_at=view.created_at,
        updated_at=view.updated_at,
    )


def _collection_to_response(view: LlmProvidersView) -> LlmProvidersResponse:
    return LlmProvidersResponse(
        providers=[_provider_to_response(p) for p in view.providers],
        default_model_profile_id=view.default_model_profile_id,
        billing_mode=view.billing_mode,
        platform_available=view.platform_available,
        platform_model=view.platform_model,
    )


@router.get("", response_model=LlmProvidersResponse)
async def list_llm_providers(
    user: AuthUser,
    service: LlmProviderService = Depends(get_llm_provider_service),
):
    """The user's BYOK providers + deployment capabilities."""
    return _collection_to_response(await service.list_providers(user.user_id))


@router.post("", response_model=LlmProviderView, status_code=201)
async def create_llm_provider(
    body: CreateLlmProviderRequest,
    user: AuthUser,
    service: LlmProviderService = Depends(get_llm_provider_service),
):
    """Add an OpenAI-compatible provider (key encrypted at rest; status 'unchecked')."""
    view = await service.create_provider(
        user.user_id,
        label=body.label,
        api_key=body.api_key,
        base_url=body.base_url,
    )
    logger.info(
        "llm_provider.created",
        user_id=user.user_id,
        provider_id=view.id,
        base_url=view.base_url,
    )
    return _provider_to_response(view)


@router.patch("/{provider_id}", response_model=LlmProviderView)
async def update_llm_provider(
    provider_id: str,
    body: UpdateLlmProviderRequest,
    user: AuthUser,
    service: LlmProviderService = Depends(get_llm_provider_service),
):
    """Update a provider (endpoint / label; key optional to keep)."""
    fields_set = set(body.model_fields_set)
    view = await service.update_provider(
        user.user_id,
        provider_id,
        label=body.label,
        api_key=body.api_key,
        base_url=body.base_url,
        fields_set=fields_set,
    )
    if "api_key" in fields_set and (body.api_key or "").strip():
        # Key material never logged — only the fact a BYOK key was rotated.
        logger.info(
            "llm_provider.key_updated",
            user_id=user.user_id,
            provider_id=provider_id,
        )
    return _provider_to_response(view)


@router.delete("/{provider_id}", response_model=StatusResponse)
async def delete_llm_provider(
    provider_id: str,
    user: AuthUser,
    service: LlmProviderService = Depends(get_llm_provider_service),
):
    """Remove a provider (profile slots referencing it fall back cleanly)."""
    await service.delete_provider(user.user_id, provider_id)
    logger.info(
        "llm_provider.deleted",
        user_id=user.user_id,
        provider_id=provider_id,
    )
    return StatusResponse()


@router.post("/{provider_id}/test", response_model=LlmProviderView)
async def test_llm_provider(
    provider_id: str,
    user: AuthUser,
    service: LlmProviderService = Depends(get_llm_provider_service),
):
    """Probe one provider's endpoint and persist 'active' / 'error' + supports_tools."""
    view = await service.test_provider(user.user_id, provider_id)
    logger.info(
        "llm_provider.tested",
        user_id=user.user_id,
        provider_id=provider_id,
        status=view.status,
        supports_tools=view.supports_tools,
        has_message=bool(view.message),
    )
    return _provider_to_response(view)
